tools/eslint-plugin/dist/rules/public-extension-exports.js

```diff
@@ -0,0 +1,222 @@
+const ts = require("typescript");
+const fs = require("fs");
+
+/** converts the numeric typescript enum value for ts.SyntaxKind to a string. Defaults to "real". */
+const getSyntaxKindFriendlyName = (syntaxKind) => {
+  const syntaxKindFriendlyNames = {
+    [ts.SyntaxKind.ClassDeclaration]: "real",
+    [ts.SyntaxKind.EnumDeclaration]: "enum",
+    [ts.SyntaxKind.InterfaceDeclaration]: "interface",
+    [ts.SyntaxKind.TypeAliasDeclaration]: "type",
+  }
+  return syntaxKindFriendlyNames[syntaxKind] || "real";
+}
+
+
+let firstRun = true;
+
+/**
+ * This rule prevents the exporting of extension APIs that not not meet certain release tags.
+ */
+module.exports = {
+  meta: {
+    type: "problem",
+    docs: {
+      description: "Prevent the exporting of extension APIs that do not meet certain release tags.",
+      category: "TypeScript",
+    },
+    messages: {
+      forbidden: `{{kind}} "{{name}}" without one of the release tags "{{releaseTags}}".`,
+    },
+    schema: [
+      {
+        type: "object",
+        additionalProperties: false,
+        properties: {
+          releaseTags: {
+            type: "array",
+            uniqueItems: true,
+            items: {
+              type: "string",
+              enum: ["public", "beta", "alpha", "internal", "preview"]
+            }
+          },
+          outputApiFile: {
+            type: "boolean"
+          }
+        }
+      }
+    ]
+  },
+
+  create(context) {
+    const parserServices = getParserServices(context);
+
+    const releaseTags = (context.options.length > 0 && context.options[0].releaseTags) || ["public"];
+    const extensionsTag = "extensions";
+    const previewTag = "preview";
+
+    const outputApiFile = (context.options.length > 0 && context.options[0].outputApiFile) || false;
+    const apiFilePath = "./lib/GeneratedExtensionApi.csv";
+
+    if (firstRun) {
+      firstRun = false;
+      if (outputApiFile) {
+        // create/clear output api file on first run
+        fs.writeFileSync(apiFilePath, "");
+      }
+    }
+
+    function addToApiList(declaration, isPreview) {
+      if (!outputApiFile) {
+        return;
+      }
+
+      const createCsvString = (name, kind) => `${name},${kind},${isPreview ? 'preview' : 'public'}\n`;
+
+      const name = declaration.kind === ts.SyntaxKind.VariableStatement ?
+        declaration.declarationList.declarations[0].symbol.escapedName :
+        declaration.symbol.escapedName;
+      const kind = getSyntaxKindFriendlyName(declaration.kind);
+      const csvString = createCsvString(name, kind);
+      fs.writeFileSync(apiFilePath, csvString, { flag: "a" });
+    }
+
+    function getFileName(parent) {
+      let currentParent = parent;
+      while (currentParent) {
+        if (currentParent.fileName !== undefined)
+          return currentParent.fileName;
+        currentParent = currentParent.parent;
+      }
+      return undefined;
+    }
+
+    function isLocalFile(declaration) {
```

Comment:
eslint is supposed to ignore node_modules by default and each file has its own AST with its own root node. So you can't go to another file by just going through parent-child relationships, at least in espree-compatible parsers. Maybe the Typescript-native AST is different, but I think you'd have to use a method in the compiler API rather than a property in the AST to jump to a different file. Have you encountered a case where you jumped over to a different file in this rule?

Reply:
Comparing `getFileName` and `context.getFilename` they always seem to be the same (in core-frontend and core-common anyways), so I think you are right! I'll remove this